### PR TITLE
perf(workspace): shard the unit-test gate across a 4-way matrix

### PR DESCRIPTION
**Pull request by ofri-peretz**
## Summary

`Unit Tests (Turbo)` is the **entire** merge-blocking critical path. The four heavy jobs run in parallel off `gate`, so the gate is only as fast as its slowest job:

| job | time |
|---|---:|
| **Unit Tests (Turbo)** | **551s** |
| Build (Turbo) | 289s |
| Typecheck (tsgo) | 98s |
| Portability Audit | 23s |

It was one `npx turbo run test` on one runner — **61 tasks, 0 cached, 8m33s**.

The `gate` job now resolves which packages turbo would actually test and splits them across a matrix. Sharding has to happen there, not inside each shard: turbo **ORs** multiple `--filter` args, so there is no way to intersect "affected since `origin/main`" with "this shard's packages" from within a shard. The gate runs `turbo run test $TURBO_FILTER_ARGS --dry=json` — already affected-narrowed — and partitions that.

Measured on the current tree: 33 packages have a real test script, split **9/8/8/8**. Executing shard 1 for real ran 27 tasks (9 test + 18 dependency builds) green in **157s** locally, against 61 tasks unsharded.

## Safety

A required check is being made to do less work per job, so the failure mode to guard is a green run that tested nothing.

- **Totality is locked.** `scripts/__tests__/shard-test-packages.test.ts` sweeps sizes `{0,1,2,7,27,34,100}` × shard counts `{1,2,3,4,6,8,12}` and asserts the union of shards equals the input exactly — no drops, no duplicates, no empty shards. 60 assertions.
- **The lock immediately earned its keep.** `Math.max(1, Math.floor(NaN))` is `NaN`, and `i % NaN` is `NaN` — so a malformed shard count silently produced **zero shards**: every test skipped, pipeline green. Now clamped to "one shard with everything".
- **Verified against real turbo output**, not just unit fixtures: the four shards resolve to `9+8+8+8 = 33` test tasks, exactly matching the unsharded baseline.
- **Empty affected set emits one no-op shard**, not an empty matrix. An empty `include` makes GitHub skip the job → `needs.test.result` is `skipped` → the aggregating gate counts that as failure. Teaching the gate to accept `skipped` instead would weaken it everywhere else.
- `set -euo pipefail` on the shard step — a turbo failure must not yield an empty matrix that skips every test and reports green.
- `fail-fast: false` so one red shard doesn't hide the others.
- Renaming the job to `Unit Tests (Turbo) <i>/<n>` is safe: branch protection requires only `oxlint (fast pass)` and `Quality (Full) Gate`, never this job directly. I checked before renaming.

## Test plan

- [x] 60/60 partitioner locks pass; the NaN case fails on the pre-fix version.
- [x] Real turbo output → 33 packages → 4 shards → union of 33, unique 33.
- [x] Shard 1 executed end-to-end: `Tasks: 27 successful, 27 total`, 157s.
- [x] `quality-full.yml` parses; **Workflow Conventions**: `✅ 32 workflow file(s) pass conventions.`
- [x] This PR's own run exercises the new gate — that's the real integration test.

## Honest accounting

This is worth roughly **40%, not 4×**. Two costs eat into it:

1. `gate` grows from ~4s to ~45s — it now needs a checkout and install to run the dry-run.
2. Each shard rebuilds its own dependency closure: 4 shards × ~18 build tasks against 34 unsharded.

Cost 2 is exactly what the **turbo cache** removes — those dependency builds become cache hits. That's the next lever and it's much bigger, but it's blocked on [#343](https://github.com/ofri-peretz/eslint/pull/343): CI is currently protected from stale-cache replays *only* by having caching switched off, so enabling it before the input-hash fix lands would import that bug straight into the gate.

Shard count is a single `SHARD_COUNT` env in the gate job — trivial to raise once the cache makes shards cheap. Public repo, so runner minutes are free; the constraint is concurrent-job slots, not cost.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

- **CI Improvements**
  - Unit tests now run in parallel across dynamically generated shards based on affected packages.
  - Builds with no affected packages skip unit testing cleanly.
  - Added automated validation for scripts and repository configuration.
  - The overall quality gate now includes script checks and reports consolidated results.

- **Tests**
  - Added comprehensive coverage for test package detection, partitioning, ordering, and edge cases.

### CODE REVIEW

File: scripts/ci/shard-test-packages.mjs [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+  // A malformed workflow input must degrade to "one shard with everything",
+  // never to "nothing".
+  const parsed = Math.floor(Number(count));
+  const n = Number.isFinite(parsed) && parsed > 0 ? parsed : 1;
+  const shards = [];
+  for (const [i, pkg] of packages.entries()) {
+    const slot = i % n;
+    (shards[slot] ??= []).push(pkg);
+  }
+  return shards.filter((s) => s && s.length > 0);
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply scripts/ci/shard-test-packages.mjs]: Applied. `partition()` now pre-sizes a dense array (`Array.from({length: Math.min(n, packages.length)})`) and indexes into it, so there is no sparse hole to guard against and the `s &&` is gone. You were right that it was unreachable — `slot = i % n` only ever writes 0…min(n, len)-1.

File: .github/workflows/quality-full.yml [hunk 2]
```diff
@@ -81,19 +85,63 @@ jobs:
...
     timeout-minutes: 15
+    strategy:
+      # Report every shard's result rather than cancelling siblings on the
+      # first failure — one red shard should not hide the state of the others.
+      fail-fast: false
+      matrix: ${{ fromJSON(needs.gate.outputs.shards) }}
     steps:
       - uses: actions/checkout@9c091bb21b7c1c1d1991bb908d89e4e9dddfe3e0 # v6
         with:
           fetch-depth: 0
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply .github/workflows/quality-full.yml]: Applied. Dropped `fetch-depth: 0` from both the test shards and the scripts-tests job — they take explicit `--filter=<pkg>` names from the matrix and never evaluate a git range. `gate` keeps full history (it runs `--filter=...[origin/main]`), as do build and typecheck. Good catch given the PR is already accounting for per-shard overhead.

File: scripts/__tests__/shard-test-packages.test.ts [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+
+        // Nothing dropped...
+        expect([...flat].sort()).toEqual([...input].sort());
+        // ...and nothing duplicated (a dupe wastes a runner but also means the
+        // balance math is lying).
+        expect(new Set(flat).size).toBe(flat.length);
+        // No empty shards — an empty shard burns a runner and setup cost.
+        expect(shards.every((s) => s.length > 0)).toBe(true);
+        // Never more shards than requested, or than there is work for.
+        expect(shards.length).toBeLessThanOrEqual(Math.min(count, Math.max(size, 0)) || 0);
```
[the remark this answers is not in the document]
Comment (ofri-peretz) [reply scripts/__tests__/shard-test-packages.test.ts]: Applied, and it was a better catch than a readability nit. The bound was satisfiable by returning *too few* shards, so it's now an equality: `expect(shards.length).toBe(Math.min(count, size))`. That also pins the `size === 0` case you flagged as vacuous under `every()` — with the equality, an empty input must yield exactly 0 shards rather than passing by virtue of an empty array.

**Comment by ofri-peretz** [comment]
Closing in favour of #356 (sharding) + #361 (the safety half of this PR).

#356 and this PR both shard the unit-test gate and both rewrite `quality-full.yml`, so only one could land. #356 is the better home for the sharding: its static `matrix: shard: [1,2,3,4]` with `ci-test-shard.mts` avoids the gate-side checkout+install (~45s) that my dynamic matrix needed, and it carries more besides.

Rather than discard the rest, I split the three things #356 does **not** contain into #361 — all independent of which sharding wins:

1. **`quality-full-gate` is skipped when `gate` fails**, and GitHub scores a skipped *required* check as passing. Latent on `main` today because `gate` only reads event metadata — but #356 gives that job a checkout, an install and a shard computation, which makes it reachable. Worth landing **before** #356.
2. **`scripts/__tests__` runs nothing but the files pinned by name.** It's outside every workspace, so `turbo run test` never reaches it — including `turbo-cache-inputs-lock.test.ts`, the lock for the #343 fix that is already on `main`.
3. **`persist-credentials: false`** on all checkouts (zizmor `artipacked`, raised as Major here).

Measurements from this branch that may be useful for #356: the unfiltered 4-way matrix ran shards at 74s / 84s / 106s / 181s against a 551s single job, taking the critical path 558s → 209s. The tail shard is the balance knob — round-robin over sorted names is a heuristic, not a packing, and there were no per-package durations available at matrix-construction time.

Thanks — the duplicate was my fault for not checking open PRs before building.
